File: Aula03/Ex06/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

// Adiciona um elemento ao final da lista
void append(Node **headRef, int value) {
    Node *newNode = (Node *)malloc(sizeof(Node));
    if (newNode == NULL) {
        fprintf(stderr, "Falha ao alocar memória.\n");
        exit(1);
    }
    newNode->data = value;
    newNode->next = NULL;

    if (*headRef == NULL) {
        *headRef = newNode;
    } else {
        Node *curr = *headRef;
        while (curr->next != NULL) {
            curr = curr->next;
        }
        curr->next = newNode;
    }
}

// Imprime a lista
void printList(Node *head) {
    Node *curr = head;
    while (curr != NULL) {
        printf("%d -> ", curr->data);
        curr = curr->next;
    }
    printf("NULL\n");
}

// Libera todos os nós da lista
void freeList(Node *head) {
    Node *curr = head;
    while (curr != NULL) {
        Node *next = curr->next;
        free(curr);
        curr = next;
    }
}

int main(void) {
    Node *head = NULL;
    int n;
    printf("Quantos elementos deseja inserir na lista? ");
    if (scanf("%d", &n) != 1 || n < 0) {
        fprintf(stderr, "Número inválido.\n");
        return 1;
    }
    for (int i = 0; i < n; i++) {
        int value;
        printf("Digite o valor %d: ", i);
        if (scanf("%d", &value) != 1) {
            fprintf(stderr, "Entrada inválida.\n");
            freeList(head);
            return 1;
        }
        append(&head, value);
    }
    printf("Lista ligada: ");
    printList(head);

    freeList(head);
    return 0;
}

File: Aula03/Ex06/main.py
class Node:
    def __init__(self, data: int) -> None:
        self.data: int = data
        self.next: 'Node | None' = None


def append(head: 'Node | None', value: int) -> Node:
    """Adiciona um nó ao final da lista e retorna a cabeça atualizada."""
    new_node = Node(value)
    if head is None:
        return new_node
    current = head
    while current.next is not None:
        current = current.next
    current.next = new_node
    return head


def print_list(head: 'Node | None') -> None:
    current = head
    while current:
        print(f"{current.data} -> ", end="")
        current = current.next
    print("None")


try:
    n = int(input("Quantos elementos deseja inserir na lista? "))
except ValueError:
    print("Número inválido.")
    raise SystemExit(1)

if n < 0:
    print("O número de elementos não pode ser negativo.")
    raise SystemExit(1)

head: Node | None = None
for i in range(n):
    while True:
        try:
            value = int(input(f"Digite o valor {i}: "))
            break
        except ValueError:
            print("Entrada inválida. Digite um número inteiro.")
    head = append(head, value)

print("Lista ligada:")
print_list(head)